django/views/generic/base.py

```diff
@@ -77,6 +78,13 @@ def view(request, *args, **kwargs):
         # and possible attributes set by decorators
         # like csrf_exempt from dispatch
         update_wrapper(view, cls.dispatch, assigned=())
```

Comment:
To my mind the real issue here is that we shouldn't be using `update_wrapper` at all (here or above in line 76). `view` is not in any way wrapped by `cls.dispatch` or `cls`. We just happen to want to achieve something similar to what we would want to achieve if we were wrapping a function.

By using `update_wrapper` we're setting the `__wrapped__` attribute, so when we call `inspect.signature` it thinks that this function is the decorator, and that the function we're actually interested in is `cls.dispatch`. That's the reason why it returns the signature from `dispatch`. 

We should just do the thing we want to do directly. Something along the lines of:
```
    for attr in functools.WRAPPER_ASSIGNMENTS: 
        try:
            value = getattr(wrapped, attr)
        except AttributeError:
            pass
        else:
            setattr(wrapper, attr, value)
    view.__dict__.update(cls.dispatch.__dict__)
```

Reply:
FWIW - These lines were committed in 2010, before python 3.2, when the `__wrapped__` attribute was added. 

Originally these lines would have worked without introducing any bugs, (but they still would have been semantically incorrect, since view isn't being wrapped).

Reply:
I was just going through the update_wrapper doc.  [Here](https://docs.python.org/3/library/functools.html#functools.update_wrapper)

It does make much more sense to directly update the view. Setting __wrapped__ does not seems the right thing. Making the update to this PR.

Reply:
Your thoughts on this @pope1ni ?

Reply:
So I agree with this in that the issue is that `inspect.signature()` by default follows `__wrapped__` which is not really what we want to happen. You can pass `follow_wrapped=False`, but we'd have to do that in lots of places and it would mess up the cases where we have genuine decorators attached to `.dispatch()`.

The solution we've implemented works because `inspect.signature()` will only follow `__wrapped__` unless it finds `__signature__`, but it is papering over the issue which is that we expect the signature of the generated view function, not to go off down the rabbit hole finding the signature of something that was wrapped.

We can avoid copying the internals of `functools.update_wrapper()` and just delete the `__wrapped__` attribute:

```python
        # Take name and docstring from the class and any possible attributes
        # set by decorators, e.g. @csrf_exempt, from the dispatch method.
        # Remove the __wrapped__ attribute to ensure that introspection returns
        # the signature of the generated view, not that of the dispatch method.
        update_wrapper(view, cls, updated=())
        update_wrapper(view, cls.dispatch, assigned=())
        del view.__wrapped__
```

This solves the problem and reduces it to a one line change (although I think it makes sense to put this all together and improve the docstring as in the snippet above - it's all related).

Reply:
I disagree with this. We have two calls to `update_wrapper` here:
 - The first call, is definitely a hack - we're not using `update_wrapper` the way it was designed to be used at all.
 - The second call is more reasonable (since `view` does actually wrap `dispatch`), but I'm still not 100% sure we should use it.

I'll expand below... 

In the first call to `update_wrapper`, what we are doing has nothing to do with wrappers; we want to borrow some attributes from the class. Incidentally, we would also want to borrow some attributes if we were wrapping a function. Just because a function (`update_wrapper`) works (for now!!!) it doesn't mean we can use that function for a purpose completely different to that for which it was intended.

**_This bug is actually a perfect example of why you shouldn't do this_**:
 - We used `update_wrapper` because it achieved the thing we wanted to achieve - even though it's not what the tool was designed for.
 - Everything worked great.
 - A perfectly reasonable, non-breaking, update was made to `update_wrapper` (based on the reasonable assumption people were using it to update wrapper functions).
 - We weren't using it for it's intended purpose, and the consequence was a bug is introduced.

We definitely shouldn't use `update_wrapper` here. What happens the next time a reasonable update is made, designed to improve the experience of using wrappers in python, but probably not appropriate here. It took 10 years for this bug to be noticed. Admittedly it's a small bug (which is probably why nobody noticed it), but the principle is an important one.

For the second call to `update_wrapper`, there is more of an argument for using it - it is at least technically updating a wrapper function. But in this case, it's still not really what `update_wrapper` was designed for (a normal decorator), and everything we actually want to do can be achieved by the one line:
```
view.__dict__.update(cls.dispatch.__dict__)
```
This is better than:
```
update_wrapper(view, cls.dispatch, assigned=())
del view.__wrapped__
```
Deleting `__wrapped__` is solving a problem we shouldn't need to solve. We should do what we want to do, not use a tool which doesn't quite fit, and then later undo all the damage done by a not-quite-perfect tool. It's better just to not do the damage in the first place.

Reply:
Ok. Let's take this a step further then.

The second call --  `update_wrapper(view, cls.dispatch, assigned=())` -- only updates `__dict__` as you've mentioned, the default value of the `updated` argument being `functools.WRAPPER_UPDATES` which is currently `('__dict__',)` in Python 3.9.

The first call -- `update_wrapper(view, cls, updated=())` -- has the default value of the `assigned` argument, `functools.WRAPPER_ASSIGNMENTS` which is `('__module__', '__name__', '__qualname__', '__doc__', '__annotations__')`. The comment in the code only mentions taking "name and docstring" from the class, but there is a test that also makes reference to `__module__`:

https://github.com/django/django/blob/54d91795408c878ad335226a87a44961d7f646a9/tests/generic_views/test_base.py#L173-L180

So we need to copy `__doc__`, `__name__`, and `__module__` to be consistent with what is currently expected. I guess `__doc__` makes sense, `__module__` is fine as that is the module that we're generating the function in, but `__name__` is questionable --- we're using the name of the class so we get something like the following:

```python
In [1]: from django.views import View                                           

In [2]: class SimpleView(View): pass 

In [3]: SimpleView.as_view()                                                    
Out[3]: <function __main__.SimpleView(request, *args, **kwargs)>
```

This doesn't look right, but I'm not sure what to do about it. A lambda function has `__name__` and `__qualname__` set to `'<lambda>'`, so perhaps we should be doing something similar here, e.g. `view.__name__ = f'{cls.__name__}.<generated>'`? This has the advantage that `repr(SimpleView.as_view())` looks like `'<function SimpleView.<generated> at 0x7f58213679d0>'` rather than `'<function SimpleView at 0x7f49d4ba0dc0>'` which is confusing as `SimpleView` is not a function but a class.

We would also need to explicitly update this to copy `__qualname__` in a similar way. That was introduced in Python 3.3 and isn't explicitly tested, so the tests would need to be updated.

The question is whether we want to copy `__annotations__`. If anything it is wrong to copy these from the class, and they should be copied from `cls.dispatch`. If we do copy it from the method, we don't need to worry about removing `self` as it isn't typically annotated. (Also, using `self` as the first argument is just a convention and not guaranteed, so we cannot assume that it should be stripped.)

@tim-mccurrach As you commented `# I'm not sure which of these we actually want` in your first comment above, what do you make of this assessment?

Assuming you're happy with that, I then propose we simplify this to:

```python
        # Take special attributes as defined on the class, and annotations and
        # any possible attributes set by decorators, e.g. @csrf_exempt, from
        # the dispatch method.
        view.__doc__ = cls.__doc__
        view.__name__ = f'{cls.__name__}.<generated>'
        view.__module__ = cls.__module__
        view.__qualname__ = f'{cls.__qualname__}.<generated>'
        view.__annotations__ = cls.dispatch.__annotations__
        view.__dict__.update(cls.dispatch.__dict__)
```

Then we also don't need to add the extra test, but can update the one I referenced above:

```python
    def test_class_attributes(self):
        """
        The callable returned from .as_view() has proper special attributes.
        """
        cls = SimpleView
        view = cls.as_view()
        self.assertEqual(view.__doc__, cls.__doc__)
        self.assertEqual(view.__name__, f'{cls.__name__}.<generated>')
        self.assertEqual(view.__module__, cls.__module__)
        self.assertEqual(view.__qualname__, f'{cls.__qualname__}.<generated>')
        self.assertEqual(view.__annotations__, cls.dispatch.__annotations__)
        self.assertFalse(hasattr(view, '__wrapped__'))
```

Reply:
It seems to be a better approach that manually removing self from the signature. Updating this PR!

Reply:
Appending . is breaking a lot of other tests. It is modifying the response.headers['X-View'] which is leading to a lot of tests to break.
Shall I refactor thoses tests as well or remove '' from the name/qualname?

Reply:
So most of the tests that are breaking with appending `.<generated>` are in the admindocs tests. These views could be updated to handle stripping the suffix appended, but I guess there are other cases where this may be incompatible with existing projects.

So drop the `.<generated>` bit for now - it was something I thought we should do based on thoroughly looking into the different attributes in my comment above. I still think using the class name directly is wrong, but perhaps that'll need more discussion. I don't know if you have any thoughts on this @adamchainz?

Reply:
I looked into why the tests are failing. It's because some internals of the URL resolver and admindocs rely on the `__name__` set here. But they could also, more accurately, use the `view_class` atttribute. Therefore I've made PR #14138 to change that.

If we go that route then I think we shouldn't even set `__name__` / `__qualname__`. Leaving them as their defaults is sensible and doesn't lie (`__name__ = 'view'` , `__qualname__ = '...View.as_view.<locals>.view'`). One can differentiate class-based views with the `view_class` attribute.

Reply:
I don't think it's worth doing fudges like that without a definite use case. For reference I checked `dataclases`  and it leaves generated functions with their standard `__qualname__`:

```
In [1]: from dataclasses import dataclass

In [2]: @dataclass
   ...: class X:
   ...:     y: int
   ...:

In [3]: X.__init__
Out[3]: <function __main__.__create_fn__.<locals>.__init__(self, y: int) -> None>
```

Reply:
Excellent. Happy with that. So, as I have above, without any `__qualname__`-mangling, should do the trick.